go-controller/pkg/ovn/dns_name_resolver/egressfirewall_external_dns_test.go

```diff
@@ -0,0 +1,663 @@
+package dnsnameresolver
+
+import (
+	"context"
+	"net"
+	"time"
+
+	"github.com/google/go-cmp/cmp"
+	"github.com/google/go-cmp/cmp/cmpopts"
+	ocpnetworkapiv1alpha1 "github.com/openshift/api/network/v1alpha1"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/config"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/factory"
+	addressset "github.com/ovn-org/ovn-kubernetes/go-controller/pkg/ovn/address_set"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/util"
+
+	"github.com/onsi/ginkgo"
+	"github.com/onsi/gomega"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+)
+
+func newDNSNameResolverObject(name, namespace, dnsName string, adddresses []string) *ocpnetworkapiv1alpha1.DNSNameResolver {
+	resolvedAddresses := []ocpnetworkapiv1alpha1.DNSNameResolverResolvedAddress{}
+	for _, address := range adddresses {
+		resolvedAddresses = append(resolvedAddresses, ocpnetworkapiv1alpha1.DNSNameResolverResolvedAddress{
+			IP: address,
+		})
+	}
+
+	return &ocpnetworkapiv1alpha1.DNSNameResolver{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: namespace,
+		},
+		Spec: ocpnetworkapiv1alpha1.DNSNameResolverSpec{
+			Name: ocpnetworkapiv1alpha1.DNSName(dnsName),
+		},
+		Status: ocpnetworkapiv1alpha1.DNSNameResolverStatus{
+			ResolvedNames: []ocpnetworkapiv1alpha1.DNSNameResolverResolvedName{
+				{
+					DNSName:           ocpnetworkapiv1alpha1.DNSName(dnsName),
+					ResolvedAddresses: resolvedAddresses,
+				},
+			},
+		},
+	}
+}
+
+var _ = ginkgo.Describe("Egress Firewall External DNS Operations", func() {
+	var (
+		extEgDNS            *ExternalEgressDNS
+		_, clusterSubnet, _ = net.ParseCIDR("10.128.0.0/14")
+		wf                  *factory.WatchFactory
+		fakeClient          *util.OVNMasterClientset
+	)
+
+	const (
+		dnsName   = "www.example.com."
+		namespace = "namespace1"
+	)
+
+	ginkgo.BeforeEach(func() {
+		config.OVNKubernetesFeature.EnableEgressFirewall = true
+		config.OVNKubernetesFeature.EnableDNSNameResolver = true
+		config.Default.ClusterSubnets = []config.CIDRNetworkEntry{{CIDR: clusterSubnet}}
+
+		fakeAddressSetFactory := addressset.NewFakeAddressSetFactory(DefaultNetworkControllerName)
+
+		fakeClient = util.GetOVNClientset().GetMasterClientset()
+		var err error
+		wf, err = factory.NewMasterWatchFactory(fakeClient)
+		gomega.Expect(err).NotTo(gomega.HaveOccurred())
+
+		extEgDNS, err = NewExternalEgressDNS(fakeAddressSetFactory, DefaultNetworkControllerName, true, wf.DNSNameResolverInformer().Informer())
+		gomega.Expect(err).NotTo(gomega.HaveOccurred())
+
+		err = wf.Start()
+		gomega.Expect(err).NotTo(gomega.HaveOccurred())
+
+		err = extEgDNS.Run()
+		gomega.Expect(err).NotTo(gomega.HaveOccurred())
+	})
+
+	ginkgo.AfterEach(func() {
+		wf.Shutdown()
+		extEgDNS.Shutdown()
+
+	})
+
+	ginkgo.Context("on dns name resolver resource creation", func() {
+		ginkgo.It("Should add IPv4 addresses", func() {
```

Comment:
looks like most test here could become tables with parameters like ipv4/ipv6/dual + ips, and the with cluster subnet/without cluster subnet